cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -370,13 +475,20 @@ class KeyHasher {
         column_arrays_[k] =
             ColumnArrayFromArrayDataAndMetadata(array_data, metadata_[k], i, length);
       }
+      // write directly to the cache
       Hashing64::HashMultiColumn(column_arrays_, &ctx_, hashes_.data() + i);
     }
-    batch_ = batch;
+#ifndef NDEBUG
+    DebugSync(*node_) << "key hasher " << index_ << " got hashes "
+                      << compute::internal::GenericToString(hashes_) << std::endl;
+#endif
```

Comment:
I wonder if it would be possible to do something like...

```
DEBUG_SYNC(*node_, "key hasher", index_, "got hashes ", compute::internal::GenericToString(hashes_), std::endl);
```

The macro could compile to nothing in release mode and then we can get rid of the `#ifndef`

Reply:
Yeah, that should work for removing those `#ifndef` used for `DebugSync`. I'll look into this.

Reply:
This is done in the [recent commit](https://github.com/apache/arrow/pull/34392/commits/85328512f4899ca92ce5404f442036f54d6ab3ee). I tried a number of other approaches that I couldn't get compile with I/O manipulator (like `std::endl`) before settling on this one.